python/sedonadb/src/udf.rs

```diff
@@ -0,0 +1,341 @@
+            let py_sedona_types = args
+                .iter()
+                .map(|arg| -> Result<_, PySedonaError> { Ok(PySedonaType::new(arg.clone())) })
+                .collect::<Result<Vec<_>, _>>()?;
+            let py_scalar_values = zip(&py_sedona_types, scalar_args)
+                .map(|(sedona_type, maybe_arg)| {
+                    maybe_arg.map(|arg| PySedonaValue {
+                        sedona_type: sedona_type.clone(),
+                        value: ColumnarValue::Scalar(arg.clone()),
+                        num_rows: 1,
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            let py_return_field =
+                self.py_return_field
+                    .call(py, (py_sedona_types, py_scalar_values), None)?;
+            if py_return_field.is_none(py) {
+                return Ok(None);
+            }
+
+            let return_type = import_sedona_type(py_return_field.bind(py))?;
+            Ok(Some(return_type))
+        })?;
+
+        Ok(return_type)
+    }
+
+    fn invoke_batch_from_args(
+        &self,
+        arg_types: &[SedonaType],
+        args: &[ColumnarValue],
+        return_type: &SedonaType,
+        num_rows: usize,
+    ) -> Result<ColumnarValue> {
+        let result = Python::with_gil(|py| -> Result<ArrayRef, PySedonaError> {
+            let py_values = zip(arg_types, args)
+                .map(|(sedona_type, arg)| PySedonaValue {
+                    sedona_type: PySedonaType::new(sedona_type.clone()),
+                    value: arg.clone(),
+                    num_rows,
+                })
+                .collect::<Vec<_>>();
+
+            let py_return_type = PySedonaType::new(return_type.clone());
+            let py_args = PyTuple::new(py, py_values)?;
+
+            let result =
+                self.py_invoke_batch
+                    .call(py, (py_args, py_return_type, num_rows), None)?;
+            let result_bound = result.bind(py);
+            if !result_bound.hasattr("__arrow_c_array__")? {
+                return Err(
+                    PySedonaError::SedonaPython(
+                        "Expected result of user-defined function to return an object implementing __arrow_c_array__()".to_string()
+                    )
+                );
+            }
+
+            let (result_field, result_array) = import_arrow_array(result_bound)?;
+            let result_sedona_type = SedonaType::from_storage_field(&result_field)?;
+
+            if return_type != &result_sedona_type {
+                let return_type_storage = SedonaType::Arrow(return_type.storage_type().clone());
+                if return_type_storage != result_sedona_type {
+                    return Err(PySedonaError::SedonaPython(format!(
+                        "Expected result of user-defined function to return array of type {return_type} or its storage but got {result_sedona_type}"
+                    )));
+                }
+            }
```

Comment:
We allow Python UDFs to return arrays with different sedona_type but the same storage type. Will this trigger some runtime type-checking failure when the result of the Python UDF is fed into some other functions?